### PR TITLE
Fixed #5618 - API v8 filtering custom field

### LINKED ISSUE #5618: API v8 filtering custom field
#### Issue

See https://suitecrm.com/suitecrm/forum/suitecrm-7-0-discussion/18407-api-filter-by-custom-field

when making a call like:
``` GET api/v8/modules/Accounts?filter[Accounts.firm_id_c]=[eq]12345678```

#### Expected Behavior

The SQL query should use accounts_cstm instead of accounts

#### Actual Behavior

suitecrm.log reports:
```
Error running count query for Account List:  Query Failed:  SELECT count(*) c FROM accounts  LEFT JOIN accounts_cstm ON accounts.id = accounts_cstm.id_c   LEFT JOIN  users jt0 ON accounts.modified_user_id=jt0.id AND jt0.deleted=0

 AND jt0.deleted=0  LEFT JOIN  users jt1 ON accounts.created_by=jt1.id AND jt1.deleted=0

 AND jt1.deleted=0  LEFT JOIN  users jt2 ON accounts.assigned_user_id=jt2.id AND jt2.deleted=0

 AND jt2.deleted=0  LEFT JOIN  accounts jt3 ON accounts.parent_id=jt3.id AND jt3.deleted=0

 AND jt3.deleted=0  LEFT JOIN  campaigns jt4 ON accounts.campaign_id=jt4.id AND jt4.deleted=0

 AND jt4.deleted=0 where (accounts.firm_id_c = 1) AND accounts.deleted=0: MySQL error 1054: Unknown column 'accounts.firm_id_c' in 'where clause'
```

#### Possible Fix

Fix FilterInterpreter::getFilterByAttributes() to check the if the field is the custom version and handle it accordingly

#### Steps to Reproduce

1. Set up api v8
2. make call`GET api/v8/modules/Accounts?filter[Accounts.firm_id_c]=`[`[eq`]`]12345678`
#### Context

high priority - there is no work around.

#### Your Environment

* SuiteCRM Version used:
* Browser name and version (e.g. Chrome Version 51.0.2704.63 (64-bit)): na
* Environment name and version (e.g. MySQL, PHP 7): PHP 5.5.9 / Maria DB / Apache
* Operating System and version (e.g Ubuntu 16.04):  Ubuntu 16.04

### PR DESCRIPTION
**Pull request by daniel-samson**
## Description

See #5618 . Developers are unable to filter on custom fields.

## Motivation and Context

Issue is due to the Filter Interpreter is not using the _cstm version of the table when a custom field has been introduced.

## How To Test This

- Set up API
- Create custom field eg. firm_id in studio
- using a rest client  make call```GET /api/v8/modules/Accounts?filter[Accounts.firm_id_c]=[eq]12345678``

** use [[ eq ]] instead of [eq] ** github is not displaying it correctly :(

You should receive a 200 status instead of a 500.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [ ] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

**Comment by brendanbell** [comment]
Hi Daniel

Just Added / Changed the following fields:

StringValidator.php
FilterInterpreter.php

I then did a rebuild in suitecrm

here is my postman call

`Accounts?filter[Accounts.firm_id_c]=[[eq]]16159010` 

Working fine.

Note:
Gotta be very careful with for formatting of this query. When I did this first time
`Accounts?filter[Accounts.firm_id_c]=[[ eq ]]16159010` 
with the spaces between [[ eq ]] it returned nothing at all, however in apache log I have an error

So perhaps some error handling is needed to throw back to user?

Here is apache error:

`PHP Fatal error:  Uncaught TypeError: Argument 3 passed to SuiteCRM\\API\\v8\\Controller\\ApiController::generateJsonApiExceptionResponse() must be an instance of Exception, instance of Error given, called in /var/www/suitecrm-development/lib/API/core/containers.php on line 119 and defined in /var/www/suitecrm-development/lib/API/v8/Controller/ApiController.php:158\nStack trace:\n#0 /var/www/suitecrm-development/lib/API/core/containers.php(119): SuiteCRM\\API\\v8\\Controller\\ApiController->generateJsonApiExceptionResponse(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Object(Error))\n#1 [internal function]: {closure}(Object(Slim\\Http\\Request), Object(Slim\\Http\\Response), Object(Error))\n#2 /var/www/suitecrm-development/vendor/slim/slim/Slim/App.php(690): call_user_func_array(Object(Closure), Array)\n#3 /var/www/suitecrm-development/vendor/slim/slim/Slim/App.php(392): Slim\\App->handlePhpError(Object(Error), Object(Slim\\Http\\Request), Object(Slim\\Http\\Response))\n#4 /var/www/suitecrm-development/vendor/slim/slim/Slim/App.php(296): Sl in /var/www/suitecrm-development/lib/API/v8/Controller/ApiController.php on line 158`

And, thank you for looking into this so quickly.

Regards
Brendan

File: lib/API/JsonApi/v1/Filters/Interpreters/FilterInterpreter.php [hunk 1]
```diff
@@ -348,4 +354,35 @@ protected function getOperator($operator)
...
+     * @param string $field
+     * @return bool
+     */
+    protected function isCustomField($field)
+    {
+        if(!is_string($field)) {
+            throw new InvalidArgumentException('isCustomField requires $field to be a string');
+        }
+
+        return StringValidator::endsWith($field, '_c');
```
Comment (PedroErnst) [inline lib/API/JsonApi/v1/Filters/Interpreters/FilterInterpreter.php]: Correct me if I'm wrong, but I think not all custom fields end with '_c' ? They probably should, but this isn't enforced. I think this function is going to miss a lot of custom fields.

**Comment by daniel-samson** [comment]
Hi @brendanbell ,

Thanks for your feedback, I have updated my pull request to include fixes for the exceptions.

Please refer to our documentation for the details of the [operators](https://docs.suitecrm.com/developer/api/version-8/json-api/#_operators)

**Comment by Dillon-Brown** [comment]
Assessed

**Pull request merged**